### PR TITLE
feat(release notes): be able to fetch release notes from Azure DevOps

### LINKED ISSUE #18324: Support changelogs from Azure DevOps
### What would you like Renovate to be able to do?

Fetch changelogs from Azure DevOps and add the changelog to the PR.

### If you have any ideas on how this should be implemented, please tell us here.

Changelog fetching is already implemented for [GitHub](https://github.com/renovatebot/renovate/tree/main/lib/workers/repository/update/pr/changelog/github) and [GitLab](https://github.com/renovatebot/renovate/tree/main/lib/workers/repository/update/pr/changelog/gitlab). So we could implement something analog for Azure DevOps using the [Azure DevOps Services REST API](https://learn.microsoft.com/en-us/rest/api/azure/devops/?view=azure-devops-rest-7.1).

Implementing this could be blocked by #14138 though.

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by sch1ldkr0ete**
## Changes

Add support for fetching release notes from Azure DevOps. (Just for markdown saved in the source repo, release lists are pretty special in Azure DevOps.)

## Context

Closes #18324

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/http/azure.spec.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+  beforeEach(() => {
+    azureApi = new AzureHttp();
+  });
+
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it('gets paginated json', async () => {
```
Comment (HonkingGoose) [inline lib/util/http/azure.spec.ts:17]:
```suggestion
  it('gets paginated JSON', async () => {
```

Abbreviations should be capitalized.

File: docs/usage/configuration-options.md [hunk 2]
```diff
@@ -798,6 +798,7 @@ Renovate can fetch release notes when they are hosted on one of these platforms:
 
 - GitHub (.com and Enterprise Server)
 - GitLab (.com and CE/EE)
+- Azure DevOps (.com)
```
Comment (viceice) [inline docs/usage/configuration-options.md:801]:
```suggestion
- Azure DevOps (.com)
- GitHub (.com and Enterprise Server)
- GitLab (.com and CE/EE)
```

sort

File: lib/workers/repository/update/pr/changelog/release-notes.spec.ts [hunk 3]
```diff
@@ -1063,6 +1119,71 @@ describe('workers/repository/update/pr/changelog/release-notes', () => {
...
+      const res = await getReleaseNotesMd(
+        {
+          ...azureProject,
+          repository: 'some-repo',
+          sourceDirectory,
+        },
+        {
+          version: '4.33.0',
+          gitRef: '4.33.0',
+        } as ChangeLogRelease
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/release-notes.spec.ts:1181]:
```suggestion
        } satisfies ChangeLogRelease
```

File: lib/workers/repository/update/pr/changelog/release-notes.spec.ts [hunk 4]
```diff
@@ -1063,6 +1119,71 @@ describe('workers/repository/update/pr/changelog/release-notes', () => {
...
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/release-notes.spec.ts:1147]:
```suggestion
        } satisfies ChangeLogRelease
```

File: lib/workers/repository/update/pr/changelog/release-notes.spec.ts [hunk 5]
```diff
@@ -1063,6 +1119,71 @@ describe('workers/repository/update/pr/changelog/release-notes', () => {
...
+      );
+
+      expect(res).toMatchSnapshot({
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/release-notes.spec.ts:1150]:
```suggestion
      expect(res).toMatchObject({
```

File: lib/workers/repository/update/pr/changelog/release-notes.spec.ts [hunk 6]
```diff
@@ -863,6 +889,36 @@ describe('workers/repository/update/pr/changelog/release-notes', () => {
...
+
+      const res = await getReleaseNotesMd(
+        {
+          ...azureProject,
+          repository: 'some-repo',
+        },
+        {
+          version: '2.0.0',
+          gitRef: '2.0.0',
+        } as ChangeLogRelease
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/release-notes.spec.ts:917]:
```suggestion
        } satisfies ChangeLogRelease
```

File: lib/workers/repository/update/pr/changelog/release-notes.spec.ts [hunk 7]
```diff
@@ -285,6 +301,16 @@ describe('workers/repository/update/pr/changelog/release-notes', () => {
         },
       ]);
     });
+
+    it('should return empty array for dev.azure.com project', async () => {
+      const res = await getReleaseList(
+        {
+          ...azureProject,
+        },
+        {} as ChangeLogRelease
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/release-notes.spec.ts:310]:
```suggestion
        {} satisfies ChangeLogRelease
```

File: lib/workers/repository/update/pr/changelog/release-notes.spec.ts [hunk 8]
```diff
@@ -57,6 +67,12 @@ const gitlabProject = {
   baseUrl: 'https://gitlab.com/',
 } as ChangeLogProject;
 
+const azureProject = {
+  type: 'azure',
+  apiBaseUrl: 'https://dev.azure.com/some-org/some-project/_apis/',
+  baseUrl: 'https://dev.azure.com/some-org/some-project/',
+} as ChangeLogProject;
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/release-notes.spec.ts:74]:
```suggestion
} satisfies ChangeLogProject;
```

File: lib/util/http/azure.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+import type { HttpOptions, HttpResponse } from './types';
+import { Http } from '.';
+
+export class AzureHttp extends Http<HttpOptions> {
+  http: any;
+  constructor(type = 'azure', options?: HttpOptions) {
+    super(type, options);
+  }
+
+  async getJsonPaginated<T>(
```
Comment (viceice) [inline lib/util/http/azure.ts:11]: Don't use new function for pagiantion. Checkout github and gitab http for pagination.

File: lib/workers/repository/update/pr/changelog/azure.spec.ts [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+    { version: '5.2.0' },
+    {
+      version: '5.4.0',
+      releaseTimestamp: '2018-08-24T14:23:00.000Z',
+    },
+    { version: '5.5.0', gitRef: 'eba303e91c930292198b2fc57040145682162a1b' },
+    { version: '5.6.0', releaseTimestamp: '2020-02-13T15:37:00.000Z' },
+    { version: '5.6.1' },
+  ],
+};
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/azure.spec.ts:28]:
```suggestion
} satisifes BranchUpgradeConfig;
```

File: lib/workers/repository/update/pr/changelog/azure.spec.ts [hunk 11]
```diff
@@ -0,0 +1,263 @@
+import * as httpMock from '../../../../../../test/http-mock';
+import * as semverVersioning from '../../../../../modules/versioning/semver';
+import * as hostRules from '../../../../../util/host-rules';
+import type { BranchUpgradeConfig } from '../../../../types';
+import { getChangeLogJSON } from '.';
+
+jest.mock('../../../../../modules/datasource/npm');
+
+const upgrade: BranchUpgradeConfig = {
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/azure.spec.ts:9]:
```suggestion
const upgrade = {
```

File: lib/workers/repository/update/pr/changelog/azure.spec.ts [hunk 12]
```diff
@@ -0,0 +1,263 @@
+import * as httpMock from '../../../../../../test/http-mock';
+import * as semverVersioning from '../../../../../modules/versioning/semver';
+import * as hostRules from '../../../../../util/host-rules';
+import type { BranchUpgradeConfig } from '../../../../types';
+import { getChangeLogJSON } from '.';
+
+jest.mock('../../../../../modules/datasource/npm');
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/azure.spec.ts:7]:
```suggestion
jest.mock('../../../../../modules/datasource');
```

better mock datasource completely

File: lib/workers/repository/update/pr/changelog/azure.spec.ts [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+        })
+      ).toBeNull();
+    });
+
+    it('works without Azure', async () => {
+      expect(
+        await getChangeLogJSON({
+          ...upgrade,
+        })
+      ).toMatchSnapshot({
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/azure.spec.ts:80]:
```suggestion
      ).toMatchObject({
```

File: lib/workers/repository/update/pr/changelog/azure.spec.ts [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+        .persist()
+        .get(
+          '/some-org//some-project/_apis/git/repositories/some-repo/items?path=/&api-version=7.0'
+        )
+        .reply(200, []);
+      expect(
+        await getChangeLogJSON({
+          ...upgrade,
+        })
+      ).toMatchSnapshot({
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/azure.spec.ts:126]:
```suggestion
      ).toMatchObject({
```

File: lib/workers/repository/update/pr/changelog/azure/index.ts [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+  AzureItem,
+  AzureTag,
+  AzureTree,
+  AzureTreeNode,
+} from '../../../../../../types/platform/azure';
+import { AzureHttp } from '../../../../../../util/http/azure';
+import { ensureTrailingSlash } from '../../../../../../util/url';
+import type { ChangeLogFile } from '../types';
+
+export const id = 'azure-changelog';
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/azure/index.ts:13]: why exporting?

File: lib/workers/repository/update/pr/changelog/azure/index.ts [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+
+export async function getTags(
+  endpoint: string,
+  repository: string
+): Promise<string[]> {
+  logger.trace('azure.getTags()');
+  const urlEncodedRepo = encodeURIComponent(repository);
+  const url = `${ensureTrailingSlash(
+    endpoint
+  )}git/repositories/${urlEncodedRepo}/refs?filter=tags&$top=100&api-version=7.0`;
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/azure/index.ts:24]: use `joinUrlParts` helper to concat url parts

File: lib/workers/repository/update/pr/changelog/azure/index.ts [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    const tags = res.body.value;
+
+    if (!tags.length) {
+      logger.debug(
+        { sourceRepo: repository },
+        'repository has no Azure DevOps tags'
+      );
+    }
+
+    return tags.map((tag) => tag.name).filter(Boolean);
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/azure/index.ts:36]:
```suggestion
    return tags.map((tag) => tag.name).filter(is.string);
```

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by viceice** [comment]
Please read our devel best practices
https://github.com/renovatebot/renovate/blob/ca66adabdb103f62722d1b87710f9e7e7158f34f/docs/development/best-practices.md#L1-L2

**Comment by sorensenmatias** [comment]
Any news on getting this merged? It would be really useful for our organization.

**Comment by sorensenmatias** [comment]
@sch1ldkr0ete @viceice Hi! Is there some way we can progress this and get it merged? It is a really important change for our organization since we deploy a lot of internal packages via Azure private feeds.

**Comment by viceice** [comment]
🤷‍♂️ it's not marked to be reviewed again. we don't monitor for changes. so the author needs to use the GitHub request review feature.

@sorensenmatias feel free to open a new PR with the requested changes to continue.

**Comment by sch1ldkr0ete** [comment]
Sorry!  This got out of my focus. I will process the requested changes this week.

**Comment by setchy** [comment]
@sch1ldkr0ete - these may be of interest. I recently refactored the changelog module https://github.com/renovatebot/renovate/pull/23076 in preparation for adding new changelog sources, eg: Bitbucket Cloud https://github.com/renovatebot/renovate/pull/22094

File: lib/workers/repository/update/pr/changelog/release-notes.spec.ts [hunk 18]
```diff
@@ -57,6 +67,12 @@
   baseUrl: 'https://gitlab.com/',
 } as ChangeLogProject;
 
+const azureProject = {
+  type: 'azure',
+  apiBaseUrl: 'https://dev.azure.com/some-org/some-project/_apis/',
+  baseUrl: 'https://dev.azure.com/some-org/some-project/',
```
[the remark this answers is not in the document]
Comment (sch1ldkr0ete) [reply lib/workers/repository/update/pr/changelog/release-notes.spec.ts:75]: Cannot see the problem here.

**Comment by rarkins** [comment]
@sch1ldkr0ete I noticed that this PR seems to still have some review comments from @viceice 

Do you expect to have time to resume work on the PR?

**Comment by rarkins** [comment]
Closing due to inactivity, PR would be welcome again

**Pull request closed without merging**